File: CSharp/SolutionLib/0201-0300/297. 二叉树的序列化与反序列化/Codec.cs
using System.Collections.Generic;
using System.Text;

namespace SolutionLib._297._二叉树的序列化与反序列化
{
    /**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
    public class Codec
    {

        // Encodes a tree to a single string.
        public string serialize(TreeNode root)
        {
            var str = new StringBuilder("[");
            var queue = new Queue<TreeNode>();
            if (root != null) queue.Enqueue(root);
            while (queue.Count != 0)
            {
                var node = queue.Dequeue();
                if (node == null)
                    str.Append("null,");
                else
                {
                    str.Append(node.val + ",");
                    queue.Enqueue(node.left);
                    queue.Enqueue(node.right);
                }
            }
            while (str.Length >= 2)
            {
                int size = str.Length;
                if (size >= 5 && str[size - 5] == 'n')
                    str.Remove(size - 5, 5);
                else
                {
                    str.Remove(str.Length - 1, 1);
                    break;
                }
            }
            str.Append("]");
            return str.ToString();
        }

        // Decodes your encoded data to tree.
        public TreeNode deserialize(string data)
        {
            data = data.Substring(1, data.Length - 2);
            var nodes = data.Split(',');
            if (nodes[0] == "") return null;
            var root = new TreeNode(int.Parse(nodes[0]));
            var queue = new Queue<TreeNode>();
            queue.Enqueue(root);
            for (int i = 1; i < nodes.Length; i += 2)
            {
                var node = queue.Dequeue();
                if (nodes[i] != "null")
                    queue.Enqueue(node.left = new TreeNode(int.Parse(nodes[i])));
                if (i + 1 < nodes.Length && nodes[i + 1] != "null")
                    queue.Enqueue(node.right = new TreeNode(int.Parse(nodes[i + 1])));
            }
            return root;
        }

    }

    // Your Codec object will be instantiated and called as such:
    // Codec ser = new Codec();
    // Codec deser = new Codec();
    // TreeNode ans = deser.deserialize(ser.serialize(root));
}

File: Java/src/SolutionLib/剑指_Offer_37/_序列化二叉树/Codec.java
package SolutionLib.剑指_Offer_37._序列化二叉树;

import UtilLib.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "[]";
        var dataList = new LinkedList<String>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            var node = queue.poll();
            if (node == null) {
                dataList.add("null");
                continue;
            }
            dataList.add(node.val + "");
            queue.offer(node.left);
            queue.offer(node.right);
        }
        while (dataList.getLast().equals("null")) {
            dataList.removeLast();
        }
        return '[' + String.join(",", dataList) + "]";
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.length() <= 2) return null;
        var split = data.substring(1, data.length() - 1).split(",");
        var root = new TreeNode(Integer.parseInt(split[0]));
        int num = split.length;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        for (int i = 1; i < num; ) {
            var node = queue.poll();
            if (!split[i].equals("null")) {
                node.left = new TreeNode(Integer.parseInt(split[i]));
                queue.offer(node.left);
            }
            i++;
            if (i < num && !split[i].equals("null")) {
                node.right = new TreeNode(Integer.parseInt(split[i]));
                queue.offer(node.right);
            }
            i++;
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));